Towards unifying symbolic and non symbolic fake tensor

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89038
* #89069
* #89074
* #89063
* #89062
* #89059

Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Towards unifying symbolic and non symbolic fake tensor

Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

anjali411 commented on test/test_proxy_tensor.py:
```diff
@@ -1337,11 +1330,15 @@ def f(a, b, c, d, e):
+}
+
 inplace_symbolic_tensor_failures = {
-    xfail('abs', ''),  # aten.abs_.default - couldn't find symbolic meta function/decomposition
```
oh because they now use python decomps instead of C++ decomps?

ezyang replied:
Specifically, I believe these started passing because previously they went through c++ composite which was not symint aware, and now they fall through to meta + fake tensor which does work. But... I didn't verify this lol

anjali411 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -325,12 +325,6 @@ Tensor empty_like(
   TensorOptions options_ = TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory);
 
-
-  TORCH_CHECK(
```
not sure why this check is being removed ...

ezyang replied:
Options constructed above is guaranteed not to have memory format

anjali411 commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1015,7 +1015,6 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
     xfail('cumsum', ''),  # aten.cumsum.default - couldn't find symbolic meta function/decomposition
     xfail('cumulative_trapezoid', ''),  # Cannot call sizes() on tensor with symbolic sizes/strides
-    xfail('deg2rad', ''),  # aten.deg2rad.default - couldn't find symbolic meta function/decomposition
     xfail('diff', ''),  # aten.zeros_like.default - couldn't find symbolic meta function/decomposition
```
shouldn't this start passing after adding the zeros_like meta impl above?

ezyang replied:
Probably hit another error

anjali411 commented on torch/_ops.py:
```diff
@@ -352,6 +352,7 @@ def handler(*args, **kwargs):
             return handler
 
         final_key = resolve_key(self, key)
+        # print(self, key, final_key)
```
not - debug comment

ezyang replied:
btw this is a SUPER useful debugging tool if you have a minimal example, it's about as useful as dispatch trace but you don't have to compile with DEBUG to get it

anjali411 replied:
haha 100% agreed.

anjali411 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -869,12 +859,6 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
             with self:
                 return func.prim_meta_impl(*args, **kwargs)
 
-        if has_symbolic_sizes:
```
why are we getting rid of this?

ezyang replied:
Well, I'm trying to unify symbolic and non-symbolic behavior. There is a major downside to dumping this, which is that the error messages are dramatically less useful here (now instead of erroring here, we'll go off into the C++ and hit an error there)

SherlockNoMad commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -821,40 +821,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                or (
+                    # TODO: Remove these exclusions, so that we can remove
+                    # this leg entirely
+                    torch_decomp_decompositions(func)
```
What remains to be done in order to remove this torch_decomp_decompositions() check?

ezyang replied:
I need to remove it and see what fails

SherlockNoMad commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -821,40 +821,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        from torch._decomp import meta_table as meta_table
 
-        # IDK: feels bad man, sym_numel on as_strided infinite loops otherwise
-        if has_symbolic_sizes and not self.cpp_meta_supports_symint(func):
```
No use of cpp_meta_supports_symint remains. 
we can remove cpp_meta_supports_symint now.

SherlockNoMad commented on torch/_meta_registrations.py:
```diff
@@ -1867,11 +1899,6 @@ def activate_meta():
-            # Attempting to register a python meta kernel for a view operator.
-            # We shouldn't do this, because the output will report as not having aliased storages.
-            # All view ops have meta kernels in C++ today, so we should use those instead.
-            pass
```
How is meta kernel working now for view ops?

ezyang replied:
I fixed the meta on prim.as_strided, so we should be indifferent to views being done in python or c++

SherlockNoMad commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -821,40 +821,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
-        # IDK: feels bad man, sym_numel on as_strided infinite loops otherwise
-        if has_symbolic_sizes and not self.cpp_meta_supports_symint(func):
-            from torch._decomp import meta_table as meta_table
+        if func not in meta_table:
```
I had always have this question? 

Why FakeTensor need to explicitly call the functions in meta_table and decomp_tables? 
Why it cannot reply on the MetaTensor under the hood, to reach the meta DispatchKey, and then call into the meta functions? 

IIUC, such path is still executed for C++ meta function at line 875. 
In meta_registration.py, we have already registered a bunch of python meta functions to the C++ dispatcher. 
So invoking meta function from C++ dispatcher should still work ?

ezyang replied:
So, to be clear, we are NOT explicitly calling `meta_table` anymore. We used to, but now we just use the meta table to try to figure out if there is a meta function that is likely to work, and if so, we'll go to the "classic" fake tensor algorithm, which is to run the meta function, and then try to figure out what the device is supposed to be afterwards. We're using the Python meta table as a proxy for "works with symbolic shapes", since Python metas automatically work with symbolic shapes. (And post your review, I readded the "cpp supports symints" as the other condition).

The question about decomp tables still stands, however. There seem to be two things going on here:

1. For the symbolic shapes case, I want to avoid hitting the C++ decomposition when there's a perfectly good Python one. But as you point out, we ought to have registered the Python decomp to the Meta key, so we'll hit it anyway. This is a good point, and I will test this.

2. Sometimes, it's better to decompose to fake tensor operations, than it is to decompose to meta tensor operations and fakeify post facto. @IvanKobzarev has been grappling with this in https://github.com/pytorch/pytorch/pull/88998 . Decomposing to fake tensor is preferable because some type promotion behavior is device sensitive, and so if you go all the way to meta you've lost important info that affects if something should pass checks or not (that being said, you can, and we in practice do, hack it up by using `device_hint` in meta registrations, so this ain't a perfect explanation). But there's also a kind of BE argument: if you have a decomposition, it ostensibly figures out device prop correctly intrinsically, so why go through the trouble of destroying the devices and then forcing fake tensor to figure it out again afterwards.

Now, what this PR does not do, is figure out when we should do (2). Empirically, this problem doesn't seem to happen with aten programs. Ivan only ran into it working with nvprims. I'm kind of thinking device_hint might be the right way to get decomps working correctly in these cases.

IvanYashchuk replied:
Currently, on master, all `torch.ops.prims` go through the meta functions with fake tensors, not meta tensors, `in_kernel_invocation_manager` is not used:
https://github.com/pytorch/pytorch/blob/f20b3f2e5734b23a9e0a898196ddf77aa90323b8/torch/_subclasses/fake_tensor.py#L864-L866

ezyang replied:
I'm going to land this PR as is, because it looks like it passes CI, and the more rigorous decomp versions don't. But I will keep running experiments to see if we can keep simplifying.

SherlockNoMad commented on torch/_prims/__init__.py:
```diff
@@ -1162,7 +1159,7 @@ def _as_strided_meta(
         )
 
-    return TensorMeta(a, shape=size, strides=stride)
+    return torch.as_strided(a, size, stride, storage_offset)
```
here...

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -869,12 +859,6 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
                 return func.prim_meta_impl(*args, **kwargs)
 
-        if has_symbolic_sizes:
-            if not self.cpp_meta_supports_symint(func):
```
i brought it back

ezyang committed:
Update on "Towards unifying symbolic and non symbolic fake tensor"


Fake tensor behaves pretty differently depending on if you have
symbolic shapes or not.  This leads to bugs; for example, we
weren't getting correct convolution_backward strides because we
bypassed the correct stride logic in fake tensor on symbolic
shapes.

This PR attempts to unify the two codepaths.  I don't manage to
unify everything, but I get most of it.  The algorithm is delicate
and I'm still hosing down test failures.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "lint only, previous ci was clean"

ezyang commented:
@pytorchbot revert -c ghfirst -m "executorch segfaults"
